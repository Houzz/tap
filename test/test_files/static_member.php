<?php

class A {
    /** @var int */
    public static $instance;

    public function testF() {
        A::$instance = 1;
        A::$instance = "abc";       // wrong
        self::$instance = 1;
        self::$instance = "abc";    // wrong
    }
}

class B {
    public static $staticVar = 1;
    public $nonStaticVar = "str";
    
    public static function staticFunc() {
        self::$staticVar = 2;
        $this->staticVar = 2;           // wrong, don't use $this in static method, and shouldn't access static via '->'
        self::$nonStaticVar = "str2";   // wrong, couldn't access non-static var from static method
        $this->nonStaticVar = "str2";   // wrong, don't use $this in static method
        self::nonStaticFunc();          // wrong, couldn't call non-static method from static method
        $this->nonStaticFunc();         // wrong, don't use $this in static method
    }

    public function nonStaticFunc() {
        self::$nonStaticVar = "str2";   // wrong, shouldn't access non-static var via '::'
        $this->nonStaticVar = "str2";
        self::$staticVar = 2;
        $this->staticVar = 3;   // wrong, should not access static variable via '->'
        self::staticFunc();
        $this->staticFunc();    // wrong
    }
}

function f() {
    B::nonStaticFunc();    // wrong
    B::staticFunc();
    $b = new B();
    $b::nonStaticFunc();    // wrong
    $b::staticFunc();
    $b->nonStaticFunc();
    $b->staticFunc();       // wrong
}

// test init static member with const
class D {
    const CONST_VALUE = 2;
    static $s1 = self::CONST_VALUE;
    static $s2 = self::CONST_VALUE2;
    const CONST_VALUE2 = 2;
    static $staticArray = array(self::CONST_VALUE, self::CONST_VALUE2);
}

__tap_check(D::$s1, "int");
__tap_check(D::$staticArray, "vector<int>");
