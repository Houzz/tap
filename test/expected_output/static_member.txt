9 E_EXPECTED_TYPE
11 E_EXPECTED_TYPE
21 E_STATIC_THIS
21 E_STATIC_PROPERTY
22 E_STATIC_PROPERTY
23 E_STATIC_THIS
24 E_STATIC_METHOD
25 E_STATIC_THIS
29 E_STATIC_PROPERTY
32 E_STATIC_PROPERTY
34 E_STATIC_METHOD
39 E_STATIC_METHOD
42 E_STATIC_METHOD
45 E_STATIC_METHOD
